proposals/distributed-async-epoch-reward.md

```diff
@@ -0,0 +1,282 @@
+nodes on the network.
+
+To address this bottleneck, we propose a new approach for distributing the
+epoch rewards over multiple blocks.
+
+
+## Detailed Design
+
+The major bottleneck for epoch reward distribution is to distribute rewards to
+stake accounts. At the time of writing, there are approximately 550K active
+stake accounts and 1.5K vote accounts on Solana Mainnet Beta. Given the
+relatively small number of vote accounts, it makes sense to keep vote rewards
+distribution mechanism unchanged. They can still be distributed efficiently at
+the first block of the epoch boundary. This reduces the impact of rewards for
+vote account and also simplifies the overall changes, and let us focus on
+solving the primary bottleneck - Stake Rewards. Only Stake rewards are going to
+be distributed out over multiple blocks.
+
+In the new stake rewards distribution approach, we will separate the
+computation of rewards from the actual distribution of rewards at the epoch
+boundary by dividing the process into two distinct phases:
+
+   1. rewards calculation phase - during which, the epoch reward for all
+      activate stake accounts are computed.
+
+   1. rewards distribution phase - during which, the calculated epoch rewards
+      for the active stake accounts are distributed.
+
+### Rewards Calculation
+
+Reward calculation phase computes all the rewards that need to be distributed
+for the active stake accounts, and partitions the reward into a number of
+chunks. Each of the chunks will be distributed by a separate block in the
+`reward distribution Phase`.
+
+To help track and verify the reward distribution in the next phase, two new
+sysvar accounts are introduced: (a) `EpochRewardHistory` sysvar account and (b)
+`RewardReserve` sysvar accounts. These sysvar accounts are updated at the
+boundary blocks (i.e. first and last blocks) in the reward calculation phase.
+More details of these two sysvar accounts are described in the following
+sections.
+
+The reward computation phase lasts *exactly* for `N` block heights since
+the beginning of the epoch. When reaching block height `N` after the start
+block of the `reward calculation phase`, the bank will mark it the end the
+`reward calculation phase`.
+
+Based on current validator statistics on Solana Mainnet Beta, a conservative
+value of 1,000 is chosen for `N`.
+
+At the end of the `reward calculation phase`, aka. `N` block height from the
+start of the epoch, the reward computation result is retrieved and partitioned
+into `M` chunks, which will be distributed during the `reward distribution`
+phase.
+
+Note that if the reward calculation finished before `N` block height, the
+result should be cached until `N` block height. If the reward calculation is
+slow and the result is not available yet at block-height `N`, the bank runtime
+will block and wait until the reward computation is finished, before processing
+any blocks with height greater than `N`.
+
```

Comment:
Maybe add a sentence before this that says something like this, as a transition:
After the full rewards list is calculated, the list is partitioned into distribution chunks.